File: catroid/src/main/java/org/catrobat/catroid/formulaeditor/FormulaInterpreter.java
/*
 * Catroid: An on-device visual programming system for Android devices
 * Copyright (C) 2010-2022 The Catrobat Team
 * (<http://developer.catrobat.org/credits>)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * An additional term exception under section 7 of the GNU Affero
 * General Public License, version 3, is available at
 * http://developer.catrobat.org/license_additional_term
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.catrobat.catroid.formulaeditor;

import org.catrobat.catroid.content.Scope;
import org.jetbrains.annotations.NotNull;

import static org.catrobat.catroid.utils.NumberFormats.trimTrailingCharacters;

public class FormulaInterpreter {

	private static final String ERROR_STRING = "ERROR";
	private final FormulaElement formulaTree;

	public FormulaInterpreter(FormulaElement formulaTree) {
		this.formulaTree = formulaTree;
	}

	public Double interpretDouble(Scope scope) throws InterpretationException {
		try {
			return assertNotNaN(interpretDoubleInternal(scope));
		} catch (ClassCastException | NumberFormatException exception) {
			throw new InterpretationException("Couldn't interpret Formula.", exception);
		}
	}

	public Integer interpretInteger(Scope scope) throws InterpretationException {
		return interpretDouble(scope).intValue();
	}

	public Float interpretFloat(Scope scope) throws InterpretationException {
		return interpretDouble(scope).floatValue();
	}

	@NotNull
	private Double interpretDoubleInternal(Scope scope) {
		Object o = formulaTree.interpretRecursive(scope);
		Double doubleReturnValue;
		if (o instanceof String) {
			doubleReturnValue = Double.valueOf((String) o);
		} else {
			doubleReturnValue = (Double) o;
		}
		return doubleReturnValue;
	}

	private double assertNotNaN(Double doubleReturnValue) throws InterpretationException {
		if (doubleReturnValue.isNaN()) {
			throw new InterpretationException("NaN in interpretDouble()");
		}
		return doubleReturnValue;
	}

	public Object interpretObject(Scope scope) {
		return formulaTree.interpretRecursive(scope);
	}

	public String interpretString(Scope scope) throws InterpretationException {
		Object interpretation = formulaTree.interpretRecursive(scope);

		if (interpretation instanceof Double && ((Double) interpretation).isNaN()) {
			throw new InterpretationException("NaN in interpretString()");
		}

		String value = String.valueOf(interpretation);
		return trimTrailingCharacters(value);
	}

	public String getUserFriendlyString(StringProvider stringProvider, Scope scope) {
		if (formulaTree.isBoolean(scope)) {
			return tryInterpretBooleanToString(stringProvider, scope);
		} else {
			return tryInterpretString(scope);
		}
	}

	private String tryInterpretString(Scope scope) {
		try {
			return interpretString(scope);
		} catch (InterpretationException interpretationException) {
			return ERROR_STRING;
		}
	}

	private String toLocalizedString(boolean value, StringProvider stringProvider) {
		return stringProvider.getTrueOrFalse(value);
	}

	public boolean interpretBoolean(Scope scope) throws InterpretationException {
		return interpretDouble(scope).intValue() != 0;
	}

	public String interpretBooleanToString(Scope scope,
			StringProvider stringProvider) throws InterpretationException {
		boolean booleanValue = interpretBoolean(scope);
		return toLocalizedString(booleanValue, stringProvider);
	}

	private String tryInterpretBooleanToString(StringProvider stringProvider, Scope scope) {
		try {
			return interpretBooleanToString(scope, stringProvider);
		} catch (InterpretationException interpretationException) {
			return ERROR_STRING;
		}
	}

	public interface StringProvider {
		String getTrueOrFalse(Boolean value);
	}
}

File: catroid/src/main/java/org/catrobat/catroid/formula/FormulaInterpreter.kt
/*
 * Catroid: An on-device visual programming system for Android devices
 * Copyright (C) 2010-2017 The Catrobat Team
 * (<http://developer.catrobat.org/credits>)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * An additional term exception under section 7 of the GNU Affero
 * General Public License, version 3, is available at
 * http://developer.catrobat.org/license_additional_term
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.catrobat.catroid.formula

import org.catrobat.catroid.formula.function.FunctionToken
import org.catrobat.catroid.formula.operator.BinaryOperatorToken
import org.catrobat.catroid.formula.operator.OperatorToken
import org.catrobat.catroid.formula.operator.UnaryOperatorToken
import org.catrobat.catroid.formula.value.ValueToken
import java.util.*

class FormulaInterpreter {

    companion object {

        fun eval(value: Boolean): Double = if (value) 1.0 else 0.0
        fun eval(value: Double): Boolean = Math.abs(value) > 0
    }

    fun eval(tokens: List<Token>): ValueToken {

        val operators = Stack<OperatorToken>()
        val values = Stack<ValueToken>()

        for (token in tokens) {
            when (token.type) {
                Token.Type.VALUE -> values.push(token as ValueToken)

                Token.Type.LEFT_BRACKET -> operators.push(token as OperatorToken)

                Token.Type.RIGHT_BRACKET -> {
                    while (!operators.empty() && operators.peek().type != Token.Type.LEFT_BRACKET) {

                        if (operators.peek() is BinaryOperatorToken) {
                            values.push((operators.pop() as BinaryOperatorToken).applyTo(values.pop(), values.pop()))
                        } else if (operators.peek() is UnaryOperatorToken) {
                            values.push((operators.pop() as UnaryOperatorToken).applyTo(values.pop()))
                        }
                    }

                    operators.pop()
                }

                Token.Type.OPERATOR -> {

                    val operator = token as OperatorToken

                    while (!operators.empty() && operators.peek().getPriority() > operator.getPriority()) {

                        if (operators.peek() is BinaryOperatorToken) {
                            values.push((operators.pop() as BinaryOperatorToken).applyTo(values.pop(), values.pop()))
                        } else if (operators.peek() is UnaryOperatorToken) {
                            values.push((operators.pop() as UnaryOperatorToken).applyTo(values.pop()))
                        }
                    }

                    operators.push(operator)
                }

                Token.Type.FUNCTION -> values.push((token as FunctionToken).eval())
            }
        }

        while (!operators.empty()) {

            if (operators.peek() is BinaryOperatorToken) {
                values.push((operators.pop() as BinaryOperatorToken).applyTo(values.pop(), values.pop()))
            } else if (operators.peek() is UnaryOperatorToken) {
                values.push((operators.pop() as UnaryOperatorToken).applyTo(values.pop()))
            }
        }

        return values.pop()
    }
}
